src/Lumper/BSP/Lumps/BspLumps/PakFileLump.cs

```diff
@@ -2,120 +2,180 @@
-        }
-        public byte[] HashMD5
-        {
-            get
+            get => _zip;
+            set
             {
-                DataStream.Seek(0, SeekOrigin.Begin);
-                return MD5.Create().ComputeHash(DataStream);
+                _zip = value;
+                UpdateEntries();
             }
         }
+
         public override void Read(BinaryReader reader, long lenght)
         {
             Stream stream = reader.BaseStream;
-            DataStream = new MemoryStream();
+            var dataStream = new MemoryStream();
             int read;
             var buffer = new byte[80 * 1024];
 
             long remaining = lenght;
             while ((read = stream.Read(buffer, 0, (int)Math.Min(buffer.Length, remaining))) > 0)
             {
-                DataStream.Write(buffer, 0, read);
+                dataStream.Write(buffer, 0, read);
                 remaining -= read;
             }
-        }
 
-        public override void Write(Stream stream)
-        {
-            if (Compress)
-                throw new InvalidDataException("Don't compress the lump containing a zip .. set the zip compression while _writing_ the zip instead");
-            DataStream.Seek(0, SeekOrigin.Begin);
-            DataStream.CopyTo(stream);
+            dataStream.Seek(0, SeekOrigin.Begin);
+            Zip = ZipArchive.Open(dataStream);
         }
 
-        public override bool Empty()
+        private void UpdateEntries()
         {
-            return DataStream is null || DataStream.Length <= 0;
+            Entries.Clear();
+            foreach (var entry in Zip.Entries)
+            {
+                Entries.Add(new PakFileEntry(entry));
+            }
         }
 
-        public PakFileLump(BspFile parent) : base(parent)
-        {
-            Compress = false;
-        }
-        public ZipArchive GetZipArchive()
+        private void UpdateZip(bool closeStream)
         {
-            DataStream.Seek(0, SeekOrigin.Begin);
-            return ZipArchive.Open(DataStream);
+            List<ZipArchiveEntry> deleteList = new();
+            foreach (var entry in Zip.Entries)
+            {
+                if (!Entries.Any(x => x.Key == entry.Key))
+                    deleteList.Add(entry);
+            }
+            foreach (var entry in deleteList)
+                Zip.RemoveEntry(entry);
+
+            foreach (var entry in Entries)
+            {
+                var existing =
+                    Zip.Entries.Where(x => x.Key == entry.Key);
+                if (existing.Any() && !entry.IsModified)
+                    continue;
+                if (existing.Count() == 1)
+                {
+                    Zip.RemoveEntry(existing.First());
+                }
+                else if (existing.Count() > 1)
+                {
+                    throw new NotImplementedException(
+                        "You can have multiple entries" +
+                        "with the same name in a zip?");
```

Comment:
Usually pakfiles are file-system based which afaik won't let you make duplicates, plus Source probably won't be happy if you have any. So even if the ZIP spec may formally support it (incl. https://learn.microsoft.com/en-us/dotnet/api/system.io.compression.zipfileextensions.createentryfromfile?view=net-7.0 remarks), I don't think we should ever support it ourselves. So a `NotSupportedException` would be more appropriate, with an error like `$"Paklump has multiple entries of the same key: {existing.Key}".`